### DISCUSSION TITLE
Failed connecting to hive metastore [localhost:9083]

### DISCUSSION THREAD

**Discussion by dominikhei** [trinodb/trino#16789] [Q&A]
I have built a local docker-compose project which involves trino, minio and a hive metastore. 

Trino is running perfectly, however when I am connected to trino via a db client and try to look at tables in the hive or iceberg catalog I always get the error:

`SQL-Fehler [16777216]: Query failed (#20230329_130607_00011_44h3f): Failed connecting to Hive metastore: [localhost:9083] Query failed (#20230329_130607_00011_44h3f): Failed connecting to Hive metastore: [localhost:9083]
Query failed (#20230329_130607_00011_44h3f): Failed connecting to Hive metastore: [localhost:9083] org.apache.thrift.TException: Failed connecting to Hive metastore: [localhost:9083]
Failed connecting to Hive metastore: [localhost:9083]
org.apache.thrift.transport.TTransportException: localhost:9083: java.net.ConnectException: Connection refused localhost:9083: java.net.ConnectException: Connection refused
org.apache.thrift.transport.TTransportException: java.net.ConnectException: Connection refused
java.net.ConnectException: Connection refused java.net.ConnectException: Connection refused Connection refused `

My docker compose file looks like the following:

```version: '3'
services:

  minio:
    image: minio/minio:RELEASE.2023-03-24T21-41-23Z
    ports:
      - "9000:9000"
      - "9001:9001"
    volumes:
      - ./minio-data:/data
    environment:
      MINIO_ACCESS_KEY: test
      MINIO_SECRET_KEY: test12334567
    command: server /data --console-address ":9001"

  createbuckets:
    image: minio/mc
    depends_on:
      - minio
    entrypoint: >
      /bin/sh -c "
      sleep 60;
      /usr/bin/mc config host add myminio http://minio:9000 test test12334567;
      /usr/bin/mc mb myminio/sample-lakehouse;
      /usr/bin/mc mb myminio/sample-lakehouse/raw;
      /usr/bin/mc mb myminio/sample-lakehouse/analytics;
      exit 0;
      "

  hive:
    image: naushadh/hive-metastore
    hostname: hive
    ports: 
      - "9083:9083"
    environment: 
      DATABASE_HOST: metastore_database
      DATABASE_DB: metastore 
      DATABASE_USER: hive
      DATABASE_PASSWORD: hive
      S3_ENDPOINT_URL: http://minio:9000
      AWS_ACCESS_KEY_ID: test
      AWS_SECRET_ACCESS_KEY: test12334567
      S3_BUCKET: sample-lakehouse
      S3_PREFIX: null
    depends_on:
      - metastore_database
      - minio 
    restart: always


  metastore_database:
    image: postgres:15
    hostname: metastore_database
    ports:
      - 5432:5432
    volumes:
      - postgres-data:/var/lib/postgresql/data
    environment:
      POSTGRES_USER: hive 
      POSTGRES_PASSWORD: hive 
      POSTGRES_DB: metastore 
  
  trino:
    image: trinodb/trino:410
    hostname: trino_container 
    ports:
      - 8080:8080 
    volumes:
      - ./trino/catalog:/etc/trino/catalog
    depends_on:
      - minio 
      - hive
  ```

The iceberg.properties file:

`connector.name=iceberg
hive.metastore.uri=thrift://hive:9083
hive.s3.endpoint=http://minio:9000
hive.s3.aws-access-key=test
hive.s3.aws-secret-key=test12334567
hive.s3.path-style-access=true
hive.s3.ssl.enabled=false `

The hive.properties is identical except for the connector name being iceberg 

Does someone know where the error could be? 

I can connect to the Postgres db using localhost but not metastore_db as a hostname from my local machine and the metastore schema is there including various tables.

**Comment by nineinchnick** [comment]
Are you sure `hive.metastore.uri` is using the `hive` hostname, and not `localhost`? This is what the error message is saying.

**Reply by dominikhei** [reply]
I changed the hive.metastore.uri in the .properties files to host.docker.internal but I got the same error: 

```
SQL-Fehler [16777216]: Query failed (#20230329_153711_00009_d743b): Failed connecting to Hive metastore: [host.docker.internal:9083]
  Query failed (#20230329_153711_00009_d743b): Failed connecting to Hive metastore: [host.docker.internal:9083]
  Query failed (#20230329_153711_00009_d743b): Failed connecting to Hive metastore: [host.docker.internal:9083]
    org.apache.thrift.TException: Failed connecting to Hive metastore: [host.docker.internal:9083]
    Failed connecting to Hive metastore: [host.docker.internal:9083]
      org.apache.thrift.transport.TTransportException: host.docker.internal:9083: java.net.ConnectException: Connection refused
      host.docker.internal:9083: java.net.ConnectException: Connection refused
        org.apache.thrift.transport.TTransportException: java.net.ConnectException: Connection refused
        java.net.ConnectException: Connection refused
          java.net.ConnectException: Connection refused
          Connection refused
```

**Comment by imvikysharma** [comment]
It looks like the error you're seeing is related to not being able to connect to the Hive Metastore. One thing to check is whether the Metastore service is actually running and accessible from within the Docker network. You can try connecting to it directly from within the hive container by running the following command:
telnet localhost 9083


If the connection is successful, then the issue may be related to how you've configured your hive container to connect to the Metastore. In particular, you may want to check the hive.metastore.uri property in your iceberg.properties file to make sure it's pointing to the correct location of the Metastore.

If the connection fails, then it's likely that the Metastore service is not running or not accessible from within the Docker network. In this case, you may want to check the logs for the hive container to see if there are any errors related to starting up the Metastore service. You may also want to try connecting to the Metastore service from another container on the same Docker network to see if the issue is specific to the hive container.

I would like to recommend a book on AWS

https://www.amazon.com/dp/B0BVJKH8PY?maas=maas_adg_AA06FAB7196A55FE3DBBEE7958CED70A_afap_abs&ref_=aa_maas&tag=maas
